File: tests/python/test_attribute_shapes.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# Public ovstage attribute-shapes test: the write→read round-trip preserves the
# three fixed/ragged column shapes — a 1-lane scalar, a fixed multi-lane tuple
# (float3, lanes in the dtype not the shape), and a ragged per-prim array. CPU-only.
# The write-flavors example is the workflow tour; this file asserts the shapes.

import numpy as np

from ovstage import DLDataType, DLDataTypeCode, OrdinalRange, PathDictionary, make_dltensor

# One float3 storage type: the per-element tuple width lives in dtype.lanes
# (a 3-lane float32), NOT in the tensor shape (which stays [prim_count]).
FLOAT3 = DLDataType(code=DLDataTypeCode.kDLFloat, bits=32, lanes=3)


def _read_one_group(stage, query, attr, end_ordinal):
    """Read a dense column written at one ordinal over the whole query: one group.
    Returns (read_handle, group); the caller releases both."""
    read = stage.read_attributes(query, [attr], OrdinalRange.latest(end_ordinal))
    read.wait()
    return read, read.fetch_next()


def test_scalar_and_fixed_lane_shapes(stage):
    with PathDictionary(stage) as paths:
        order = ["/World/A", "/World/B", "/World/C"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            # [snippet:attribute-shapes-fixed]
            # A fixed-size column stacks one row per prim in ONE tensor: a scalar
            # is a 1-lane dtype; a float3 is the SAME write with a 3-lane dtype.
            # The shape stays [prim_count]; the tuple width lives in dtype.lanes.
            temperature = paths.intern_token("temperature")
            velocity = paths.intern_token("velocity")
            stage.write_attribute(
                query, temperature, ordinal=1,
                tensors=np.array([18.5, 19.5, 20.5], np.float32), is_array=False,
            ).wait()
            stage.write_attribute(
                query, velocity, ordinal=1,
                tensors=make_dltensor(
                    np.array([1, 0, 0, 0, 1, 0, 0, 0, 1], np.float32), dtype=FLOAT3, shape=[3]
                ),
                is_array=False,
            ).wait()
            stage.advance_write_floor(ordinal=1).wait()
            # [/snippet:attribute-shapes-fixed]

            read, group = _read_one_group(stage, query, temperature, 1)
            assert group is not None
            scalar = group.array(0)
            assert group.tensor(0).dtype.lanes == 1
            assert bool(np.allclose(scalar, [18.5, 19.5, 20.5]))
            stage.release_group(group)
            read.release().wait()

            read, group = _read_one_group(stage, query, velocity, 1)
            assert group is not None
            # A 3-lane column reads back as a flat view of prim_count * 3 values.
            vec = np.asarray(group.array(0)).reshape(-1)
            assert group.tensor(0).dtype.lanes == 3
            assert vec.shape[0] == 9
            assert bool(np.allclose(vec, [1, 0, 0, 0, 1, 0, 0, 0, 1]))
            stage.release_group(group)
            read.release().wait()
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)


def test_ragged_array_shape(stage):
    with PathDictionary(stage) as paths:
        order = ["/World/A", "/World/B", "/World/C"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            # [snippet:attribute-shapes-ragged]
            # is_array=True declares a ragged (variable-length per prim) column —
            # never inferred from the payload. A list of tensors carries one row
            # per prim (lengths 2 / 3 / 1 here); read groups mirror that with one
            # tensor per prim.
            samples = paths.intern_token("samples")
            rows = [
                np.array([1.0, 2.0], np.float32),
                np.array([3.0, 4.0, 5.0], np.float32),
                np.array([6.0], np.float32),
            ]
            stage.write_attribute(query, samples, ordinal=1, tensors=rows, is_array=True).wait()
            stage.advance_write_floor(ordinal=1).wait()
            # [/snippet:attribute-shapes-ragged]

            read, group = _read_one_group(stage, query, samples, 1)
            assert group is not None
            got = {}
            for local in range(group.prim_count):
                path = order[group.prim_index(local)]
                row = group.data_row_index(local) if group.has_data_index_map else local
                got[path] = [float(v) for v in np.asarray(group.array(row))]
            stage.release_group(group)
            read.release().wait()

            assert got == {"/World/A": [1.0, 2.0], "/World/B": [3.0, 4.0, 5.0], "/World/C": [6.0]}
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)

File: tests/c/test_attribute_shapes.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// Public ovstage attribute-shapes test: the write→read round-trip preserves the
// three fixed/ragged column shapes — a 1-lane scalar, a fixed multi-lane tuple
// (float3, lanes in the dtype not the shape), and a ragged per-prim array.
// CPU-only. The write-flavors example is the workflow tour; this file asserts it.

#include <ovstage/ovstage.h>
#include <ovx/path_dictionary/path_dictionary.h>
#include <ovx/path_dictionary/path_dictionary_utils.h>
#include <dlpack/dlpack.h>

#include <cstdint>
#include <cstring>
#include <map>
#include <vector>

#include <gtest/gtest.h>

namespace
{

ovx_string_t str(const char* s)
{
    return ovx_string_t{ s, std::strlen(s) };
}

bool waitOp(ovstage_instance_t* stage, ovstage_enqueue_result_t enq, const char* what)
{
    if (enq.status != OVSTAGE_OK)
    {
        ADD_FAILURE() << what << " enqueue rejected (code " << enq.status << ")";
        return false;
    }
    ovstage_op_wait_result_t wait{};
    const ovstage_api_status_t werr = ovstage_wait_op(stage, enq.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    for (size_t i = 0; i < wait.error_op_id_count; ++i)
    {
        const ovx_string_t e = ovstage_get_last_op_error(stage, wait.error_op_ids[i]);
        ADD_FAILURE() << what << " op failed: " << std::string(e.ptr ? e.ptr : "", e.ptr ? e.length : 0);
    }
    ovstage_release_op(stage, enq.op_index);
    if (werr != OVSTAGE_OK)
        ADD_FAILURE() << what << " wait failed (code " << werr << ")";
    return werr == OVSTAGE_OK && wait.error_op_id_count == 0;
}

class ShapesTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        ovstage_instance_desc_t desc{};
        desc.name = "test.ovstage.shapes";
        ASSERT_EQ(ovstage_create_instance(&desc, &stage_), OVSTAGE_OK);
        dict_ = ovstage_get_path_dictionary(stage_);
        ASSERT_NE(dict_, nullptr);

        const ovx_string_t paths[] = { str("/World/A"), str("/World/B"), str("/World/C") };
        ASSERT_EQ(path_dictionary_create_path_list_from_strings(dict_, paths, 3, &pathList_).status, OVX_API_SUCCESS);
        ASSERT_EQ(ovstage_query_from_path_list(stage_, pathList_, &query_), OVSTAGE_OK);
    }

    void TearDown() override
    {
        if (stage_ && query_ != OVSTAGE_INVALID_QUERY_HANDLE)
            waitOp(stage_, ovstage_release_query(stage_, query_), "release_query");
        if (dict_ && pathList_ != OVX_INVALID_PRIMPATH_LIST)
            path_dictionary_release_path_list_reference(dict_, pathList_);
        if (stage_)
            ovstage_destroy_instance(stage_);
    }

    ovx_token_t intern(const char* name)
    {
        const ovx_string_t s = str(name);
        ovx_token_t token = OVX_INVALID_TOKEN;
        EXPECT_EQ(path_dictionary_create_tokens_from_strings(dict_, &s, 1, &token).status, OVX_API_SUCCESS);
        return token;
    }

    void seal(ovstage_ordinal_t ordinal)
    {
        ovstage_write_floor_desc_t floor{};
        floor.ordinal = ordinal;
        floor.scope = OVSTAGE_SCOPE_ALL;
        ASSERT_TRUE(waitOp(stage_, ovstage_advance_write_floor(stage_, &floor), "advance_write_floor"));
    }

    ovstage_instance_t* stage_ = nullptr;
    path_dictionary_instance_t* dict_ = nullptr;
    ovx_primpath_list_t pathList_ = OVX_INVALID_PRIMPATH_LIST;
    ovstage_query_handle_t query_ = OVSTAGE_INVALID_QUERY_HANDLE;
};

TEST_F(ShapesTest, ScalarAndFixedLane)
{
    const ovx_token_t temperature = intern("temperature");
    const ovx_token_t velocity = intern("velocity");

    // [snippet:attribute-shapes-fixed-c]
    // A fixed-size column stacks one row per prim in ONE tensor: a scalar is a
    // 1-lane dtype; a float3 is the SAME write with a 3-lane dtype. The shape
    // stays {prim_count}; the tuple width lives in dtype.lanes.
    float temperatures[] = { 18.5f, 19.5f, 20.5f };
    float velocities[] = { 1, 0, 0, 0, 1, 0, 0, 0, 1 }; // 3 prims x 3 lanes
    int64_t rows[] = { 3 };
    int64_t strides[] = { 1 };

    DLTensor tempTensor{};
    tempTensor.data = temperatures;
    tempTensor.device = { kDLCPU, 0 };
    tempTensor.ndim = 1;
    tempTensor.dtype = { kDLFloat, 32, 1 };
    tempTensor.shape = rows;
    tempTensor.strides = strides;

    DLTensor velTensor{};
    velTensor.data = velocities;
    velTensor.device = { kDLCPU, 0 };
    velTensor.ndim = 1;
    velTensor.dtype = { kDLFloat, 32, 3 };
    velTensor.shape = rows;
    velTensor.strides = strides;

    ovstage_write_data_t scalarWrite{};
    scalarWrite.tensors = &tempTensor;
    scalarWrite.tensor_count = 1;
    scalarWrite.is_array = false;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { temperature, {} }, 1, scalarWrite, OVSTAGE_PRIM_MODE_UPSERT),
        "write scalar"));

    ovstage_write_data_t vecWrite{};
    vecWrite.tensors = &velTensor;
    vecWrite.tensor_count = 1;
    vecWrite.is_array = false;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { velocity, {} }, 1, vecWrite, OVSTAGE_PRIM_MODE_UPSERT),
        "write vec3"));
    seal(1);
    // [/snippet:attribute-shapes-fixed-c]

    // Read the scalar column: one group, 3 rows, 1 lane.
    {
        ovstage_ordinal_range_t range{};
        range.end_ordinal = 1;
        ovstage_read_handle_t read = OVSTAGE_INVALID_READ_HANDLE;
        ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query_, &temperature, 1, range, &read),
            "read scalar"));
        ovstage_read_group_t group{};
        ASSERT_EQ(ovstage_fetch_read_next(stage_, read, OVSTAGE_TIMEOUT_INFINITE, &group), OVSTAGE_OK);
        ASSERT_EQ(group.data.tensor_count, 1u);
        EXPECT_EQ(group.data.tensors[0].dtype.lanes, 1u);
        float out[3] = {};
        std::memcpy(out, group.data.tensors[0].data, sizeof(out));
        EXPECT_FLOAT_EQ(out[0], 18.5f);
        EXPECT_FLOAT_EQ(out[1], 19.5f);
        EXPECT_FLOAT_EQ(out[2], 20.5f);
        ovstage_release_group(stage_, &group);
        waitOp(stage_, ovstage_release_read(stage_, read), "release_read");
    }

    // Read the fixed-lane column: one group, 3 rows x 3 lanes = 9 flat floats.
    {
        ovstage_ordinal_range_t range{};
        range.end_ordinal = 1;
        ovstage_read_handle_t read = OVSTAGE_INVALID_READ_HANDLE;
        ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query_, &velocity, 1, range, &read),
            "read vec3"));
        ovstage_read_group_t group{};
        ASSERT_EQ(ovstage_fetch_read_next(stage_, read, OVSTAGE_TIMEOUT_INFINITE, &group), OVSTAGE_OK);
        ASSERT_EQ(group.data.tensor_count, 1u);
        EXPECT_EQ(group.data.tensors[0].dtype.lanes, 3u);
        float out[9] = {};
        std::memcpy(out, group.data.tensors[0].data, sizeof(out));
        const float expected[9] = { 1, 0, 0, 0, 1, 0, 0, 0, 1 };
        for (int i = 0; i < 9; ++i)
            EXPECT_FLOAT_EQ(out[i], expected[i]);
        ovstage_release_group(stage_, &group);
        waitOp(stage_, ovstage_release_read(stage_, read), "release_read");
    }
}

TEST_F(ShapesTest, RaggedArray)
{
    const ovx_token_t samples = intern("samples");

    // [snippet:attribute-shapes-ragged-c]
    // is_array=true declares a ragged (variable-length per prim) column — never
    // inferred from the payload. With tensor_count > 1 each tensor is one prim's
    // row, so per-prim lengths may differ (2 / 3 / 1 here). Read groups mirror
    // that transport: one tensor per covered prim.
    float rowA[] = { 1.0f, 2.0f };
    float rowB[] = { 3.0f, 4.0f, 5.0f };
    float rowC[] = { 6.0f };
    int64_t lenA[] = { 2 }, lenB[] = { 3 }, lenC[] = { 1 };
    int64_t strides[] = { 1 };

    auto makeRow = [&](float* data, int64_t* shape) {
        DLTensor t{};
        t.data = data;
        t.device = { kDLCPU, 0 };
        t.ndim = 1;
        t.dtype = { kDLFloat, 32, 1 };
        t.shape = shape;
        t.strides = strides;
        return t;
    };
    DLTensor rowTensors[] = { makeRow(rowA, lenA), makeRow(rowB, lenB), makeRow(rowC, lenC) };

    ovstage_write_data_t raggedWrite{};
    raggedWrite.tensors = rowTensors;
    raggedWrite.tensor_count = 3;
    raggedWrite.is_array = true;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { samples, {} }, 1, raggedWrite, OVSTAGE_PRIM_MODE_UPSERT),
        "write ragged"));
    seal(1);
    // [/snippet:attribute-shapes-ragged-c]

    ovstage_ordinal_range_t range{};
    range.end_ordinal = 1;
    ovstage_read_handle_t read = OVSTAGE_INVALID_READ_HANDLE;
    ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query_, &samples, 1, range, &read), "read ragged"));

    // Collect each covered prim's row keyed by length (order-independent): the
    // three ragged rows are 2/3/1 elements long. A latest read may arrive across
    // several groups, so iterate to END_OF_ITERATION.
    std::map<int64_t, std::vector<float>> byLength;
    for (;;)
    {
        ovstage_read_group_t group{};
        const ovstage_api_status_t fetched = ovstage_fetch_read_next(stage_, read, OVSTAGE_TIMEOUT_INFINITE, &group);
        if (fetched == OVSTAGE_ERROR_END_OF_ITERATION)
            break;
        ASSERT_EQ(fetched, OVSTAGE_OK);
        for (uint32_t local = 0; local < group.prims.count; ++local)
        {
            const uint32_t rowIdx = group.data.index_map ? group.data.index_map[local] : local;
            const DLTensor& t = group.data.tensors[rowIdx];
            byLength[t.shape[0]] = std::vector<float>(static_cast<const float*>(t.data),
                                                      static_cast<const float*>(t.data) + t.shape[0]);
        }
        ovstage_release_group(stage_, &group);
    }
    waitOp(stage_, ovstage_release_read(stage_, read), "release_read");

    ASSERT_EQ(byLength.size(), 3u);
    EXPECT_EQ(byLength[2], (std::vector<float>{ 1.0f, 2.0f }));
    EXPECT_EQ(byLength[3], (std::vector<float>{ 3.0f, 4.0f, 5.0f }));
    EXPECT_EQ(byLength[1], (std::vector<float>{ 6.0f }));
}

} // namespace
